usr/src/man/man1/mt.1
'\" te
.\"  Copyright (c) 2007, Sun Microsystems, Inc. All Rights Reserved.
.\" The contents of this file are subject to the terms of the Common Development and Distribution License (the "License").  You may not use this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or http://www.opensolaris.org/os/licensing.  See the License for the specific language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable, add the following below this CDDL HEADER, with the fields enclosed by brackets "[]" replaced with your own identifying information: Portions Copyright [yyyy] [name of copyright owner]
.TH MT 1 "Jun 21, 2007"
.SH NAME
mt \- magnetic tape control
.SH SYNOPSIS
.LP
.nf
\fBmt\fR [\fB-f\fR \fItapename\fR] \fIcommand\fR... [\fIcount\fR]
.fi

.SH DESCRIPTION
.LP
The \fBmt\fR utility sends commands to a magnetic tape drive. If \fB-f\fR
\fItapename\fR is not specified, the environment variable \fBTAPE\fR is used.
If \fBTAPE\fR does not exist, \fBmt\fR uses the device \fB/dev/rmt/0n\fR.
.SH OPTIONS
.LP
The following options are supported:
.sp
.ne 2
.na
\fB\fB-f\fR \fItapename\fR\fR
.ad
.RS 15n
Specifies the raw tape device.
.RE

.SH OPERANDS
.LP
The following operands are supported:
.sp
.ne 2
.na
\fB\fIcount\fR\fR
.ad
.RS 11n
The number of times that the requested operation is to be performed. By
default, \fBmt\fR performs \fIcommand\fR once. Multiple operations of
\fIcommand\fR can be performed by specifying \fIcount\fR.
.RE

.sp
.ne 2
.na
\fB\fIcommand\fR\fR
.ad
.RS 11n
The following available commands that can be sent to a magnetic tape drive are
supported. Only as many characters as are required to uniquely identify a
\fIcommand\fR need be specified.
.sp
.ne 2
.na
\fB\fBasf\fR\fR
.ad
.RS 10n
Specifies absolute space to \fIcount\fR file number. This is equivalent to a
\fBrewind\fR followed by a \fBfsf\fR \fIcount\fR.
.RE

.sp
.ne 2
.na
\fB\fBbsf\fR\fR
.ad
.RS 10n
Back spaces over \fIcount\fR EOF marks. The tape is positioned on the
beginning-of-tape side of the EOF mark.
.RE

.sp
.ne 2
.na
\fB\fBbsr\fR\fR
.ad
.RS 10n
Back spaces \fIcount\fR records.
.RE

.sp
.ne 2
.na
\fB\fBbssf\fR\fR
.ad
.RS 10n
Back spaces over the requested number of sequential file marks. Sequential file
marks are where the file marks are one right after the other with no other
blocks of any kind between the file marks. The number argument specifies how
many sequential file marks to which to space. For example, \fBbssf 4\fR
searches backwards to the first place where there are 4 sequential file marks
and positions to the BOP side of the 4th file mark.
.sp
This command is not supported by all drives.
.RE

.sp
.ne 2
.na
\fB\fBeof\fR\fR
.ad
.br
.na
\fB\fBweof\fR\fR
.ad
.RS 10n
Writes \fIcount\fR EOF marks at the current position on the tape.
.RE

.sp
.ne 2
.na
\fB\fBfsf\fR\fR
.ad
.RS 10n
Forward spaces over \fIcount\fR EOF marks. The tape is positioned on the first
block of the file.
.RE

.sp
.ne 2
.na
\fB\fBfsr\fR\fR
.ad
.RS 10n
Forward spaces \fIcount\fR records.
.RE

.sp
.ne 2
.na
\fB\fBfssf\fR\fR
.ad
.RS 10n
Forward spaces the over requested number of sequential file marks. Sequential
file marks are where the file marks are one right after the other with no other
blocks of any kind between the file marks. The number argument specifies how
many sequential file marks to which to space. For example, \fBfssf 4\fR
searches forwards to the first place where there are 4 sequential file marks
and positions after the 4th file mark.
.sp
This command is not supported by all drives.
.RE

.sp
.ne 2
.na
\fB\fBload\fR\fR
.ad
.RS 10n
Requests drive load and thread current media. Not supported by all drives.
.RE

.sp
.ne 2
.na
\fB\fBlock\fR\fR
.ad
.RS 10n
Prevents media removal.
.RE

.sp
.ne 2
.na
\fB\fBnbsf\fR\fR
.ad
.RS 10n
Back spaces \fIcount\fR files. The tape is positioned on the first block of the
file. This is equivalent to \fIcount+1\fR \fBbsf\fRs followed by one \fBfsf\fR.
.RE

.sp
.ne 2
.na
\fB\fBseek\fR\fR
.ad
.RS 10n
Positions to requested logical tape position.
.RE

.sp
.ne 2
.na
\fB\fBtell\fR\fR
.ad
.RS 10n
Gets and prints current logical tape position.
.RE

.sp
.ne 2
.na
\fB\fBunlock\fR\fR
.ad
.RS 10n
Allows media removal.
.RE

If \fIcount\fR is specified with any of the following commands, the \fIcount\fR
is ignored and the command is performed only once.
.sp
.ne 2
.na
\fB\fBconfig\fR\fR
.ad
.RS 16n
Reads the drives current configuration from the driver and displays it in
\fBst.conf\fR format. See \fBst\fR(4D) for definition of fields and there
meanings.
.RE

.sp
.ne 2
.na
\fB\fBeom\fR\fR
.ad
.RS 16n
Spaces to the end of recorded media on the tape. This is useful for appending
files onto previously written tapes.
.RE

.sp
.ne 2
.na
\fB\fBerase\fR\fR
.ad
.RS 16n
Erases the entire tape.
.sp
Some tape drives have option settings where only portions of the tape can be
erased. Be sure to select the correct setting to erase the whole tape. Erasing
a tape can take a long time depending on the device and/or tape. Refer to the
device specific manual for time details.
.RE

.sp
.ne 2
.na
\fB\fBforcereserve\fR\fR
.ad
.RS 16n
Attempts to break a SCSI II reserve issued by another initiator. When this
command completes, the drive is not reserved for the current initiator, but is
available for use. This command can be only be executed by those with
super-user privileges.
.RE

.sp
.ne 2
.na
\fB\fBoffline\fR\fR
.ad
.br
.na
\fB\fBrewoffl\fR\fR
.ad
.RS 16n
Rewinds the tape and, if appropriate, takes the drive unit off-line by
unloading the tape.
.RE

.sp
.ne 2
.na
\fB\fBrelease\fR\fR
.ad
.RS 16n
Re-establishes the default behavior of releasing at close.
.RE

.sp
.ne 2
.na
\fB\fBreserve\fR\fR
.ad
.RS 16n
Allows the tape drive to remain reserved after closing the device. The drive
must then be explicitly released.
.RE

.sp
.ne 2
.na
\fB\fBretension\fR\fR
.ad
.RS 16n
Rewinds the cartridge tape completely, then winds it forward to the end of the
reel and back to beginning-of-tape to smooth out tape tension.
.RE

.sp
.ne 2
.na
\fB\fBrewind\fR\fR
.ad
.RS 16n
Rewinds the tape.
.RE

.sp
.ne 2
.na
\fB\fBstatus\fR\fR
.ad
.RS 16n
Prints status information about the tape unit.
.sp
Status information can include the sense key reported by the drive, the
residual and retries for the last operation, the current tape position reported
in file number, and the number of blocks from the beginning of that file. It
might also report that WORM media is loaded in that drive.
.RE

.RE

.SH EXIT STATUS
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 5n
All operations were successful.
.RE

.sp
.ne 2
.na
\fB\fB1\fR\fR
.ad
.RS 5n
Command was unrecognized or \fBmt\fR was unable to open the specified tape
drive.
.RE

.sp
.ne 2
.na
\fB\fB2\fR\fR
.ad
.RS 5n
An operation failed.
.RE

.SH FILES
.ne 2
.na
\fB\fB/dev/rmt/*\fR\fR
.ad
.RS 14n
magnetic tape interface
.RE

.SH SEE ALSO
.LP
.BR tar (1),
.BR tcopy (1),
.BR ar.h (3HEAD),
.BR st (4D),
.BR mtio (4I),
.BR attributes (7)
.SH BUGS
.LP
Not all devices support all options. Some options are hardware-dependent. Refer
to the corresponding device manual page.
.sp
.LP
\fBmt\fR is architecture sensitive. Heterogeneous operation (that is, SPARC to
x86 or the reverse) is not supported.


usr/src/cmd/mt/mt.c
/*
 * Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 */

/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley Software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#pragma ident	"%Z%%M%	%I%	%E% SMI"

/*
 * mt -- magnetic tape manipulation program
 */
#include <stdio.h>
#include <ctype.h>

#include <errno.h>
#include <sys/types.h>
#include <sys/mtio.h>
#include <sys/ioctl.h>
#include <sys/param.h>
#include <sys/buf.h>
#include <sys/conf.h>
#include <sys/file.h>
#include <sys/uio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/stat.h>
#include <sys/scsi/targets/stdef.h>
#include <fcntl.h>


#define	equal(s1, s2)	(strcmp(s1, s2) == 0)
#define	MTASF		100	/* absolute file positioning; first file is 0 */

/*
 * This can't be DEFTAPE in mtio.h because that is currently the rewinding
 * unit which makes 'mt fsf' a questionable activity at best.
 */
#define	DEFAULT_NRW_TAPE	"/dev/rmt/0n"

static int print_config(int mtfd);
static char *print_key(short key_code);
static void printreg(char *, ushort_t, char *);
static int status(int mtfd, struct mtget *);

/* Pseudo flag for open even if drive is not ready (Unloaded) or reserved */
#define	O_UNLOAD (O_RDWR | O_NDELAY)

static const struct commands {
	char *c_name;
	int c_code;
	int c_oflag;
	int c_usecnt;
} com[] = {
	{ "weof",		MTWEOF,			O_RDWR,		1 },
	{ "eof",		MTWEOF,			O_RDWR,		1 },
	{ "fsf",		MTFSF,			O_RDONLY,	1 },
	{ "bsf",		MTBSF,			O_RDONLY,	1 },
	{ "asf",		MTASF,			O_RDONLY,	1 },
	{ "fsr",		MTFSR,			O_RDONLY,	1 },
	{ "bsr",		MTBSR,			O_RDONLY,	1 },
	{ "rewind",		MTREW,			O_RDONLY,	0 },
	{ "offline",		MTOFFL,			O_RDONLY,	0 },
	{ "rewoffl",		MTOFFL,			O_RDONLY,	0 },
	{ "status",		MTNOP,			O_RDONLY,	0 },
	{ "retension",		MTRETEN,		O_RDONLY,	0 },
	{ "erase",		MTERASE,		O_RDWR,		0 },
	{ "eom",		MTEOM,			O_RDONLY,	0 },
	{ "nbsf",		MTNBSF,			O_RDONLY,	1 },
	{ "reserve",		MTIOCRESERVE,		O_RDONLY,	0 },
	{ "release",		MTIOCRELEASE,		O_RDONLY,	0 },
	{ "forcereserve",	MTIOCFORCERESERVE,	O_UNLOAD,	0 },
	{ "config",		MTIOCGETDRIVETYPE,	O_UNLOAD,	0 },
	{ "fssf",		MTFSSF,			O_RDONLY,	1 },
	{ "bssf",		MTBSSF,			O_RDONLY,	1 },
	{ "tell",		MTTELL,			O_RDONLY,	0 },
	{ "seek",		MTSEEK,			O_RDONLY,	1 },
	{ "load",		MTLOAD,			O_UNLOAD,	0 },
	{ "lock",		MTLOCK,			O_RDONLY,	0 },
	{ "unlock",		MTUNLOCK,		O_RDONLY,	0 },
	{ 0 }
};


int
main(int argc, char **argv)
{
	char *cp;
	char *tape;
	int mtfd;
	struct commands const *comp;
	struct mtget mt_status;
	struct mtlop mt_com;

	if (argc > 2 && (equal(argv[1], "-t") || equal(argv[1], "-f"))) {
		argc -= 2;
		tape = argv[2];
		argv += 2;
	} else {
		tape = getenv("TAPE");
		if (tape == NULL) {
			tape = DEFAULT_NRW_TAPE;
		}
	}

	if (argc < 2) {
		(void) fprintf(stderr,
		    "usage: mt [ -f device ] command [ count ]\n");
		return (1);
	}

	cp = argv[1];
	for (comp = com; comp->c_name != NULL; comp++) {
		if (strncmp(cp, comp->c_name, strlen(cp)) == 0) {
			break;
		}
	}

	if (comp->c_name == NULL) {
		(void) fprintf(stderr, "mt: unknown command: %s\n", cp);
		return (1);
	}

	mtfd = open(tape, comp->c_oflag);
	if (mtfd < 0) {

		/*
		 * Provide additional error message decoding since
		 * we need additional error codes to fix them problem.
		 */
		if (errno == EIO) {
			(void) fprintf(stderr,
			    "%s: no tape loaded or drive offline\n", tape);
		} else if (errno == EACCES) {
			(void) fprintf(stderr,
			    "%s: write protected or reserved.\n", tape);
		} else {
			perror(tape);
		}
		return (1);
	}

	if (comp->c_code == MTIOCFORCERESERVE ||
	    comp->c_code == MTIOCRESERVE ||
	    comp->c_code == MTIOCRELEASE)  {
		/*
		 * Handle all MTIOC ioctls used in
		 * reservation/release/takeownership.
		 */
		if (ioctl(mtfd, comp->c_code) < 0) {
			perror("mt");
			return (2);
		}
	} else if (comp->c_code == MTASF) {
		/*
		 * Handle absolute file positioning.  Ask tape driver
		 * where tape is and then skip to desired file.  If
		 * driver doesn't support get location ioctl, rewind
		 * the tape and then space to the desired file.
		 */
		int usecnt;
		daddr_t mt_fileno;

		usecnt = argc > 2 && comp->c_usecnt;
		mt_fileno = usecnt ? atol(argv[2]) : 1;
		if (mt_fileno < 0) {
			(void) fprintf(stderr, "mt: negative file number\n");
			return (1);
		}
		(void) ioctl(mtfd, MTIOCGET, (char *)&mt_status);
		if (ioctl(mtfd, MTIOCGET, (char *)&mt_status) < 0) {
			perror("mt");
			return (2);
		}
		/*
		 * Check if device supports reporting current file
		 * tape file position.  If not, rewind the tape, and
		 * space forward.
		 *
		 * If file number is -1 tape position is unknown!
		 */
		if ((mt_status.mt_flags & MTF_ASF) == 0 ||
		    (mt_status.mt_fileno == -1)) {
			/* printf("mt: rewind\n"); */
			mt_com.mt_count = 1;
			mt_com.mt_op = MTREW;
			if (ioctl(mtfd, MTIOCLTOP, &mt_com) < 0) {
				(void) fprintf(stderr, "%s %s %ld ",
				    tape, comp->c_name, mt_fileno);
				perror("mt");
				return (2);
			}
			/* Needed to rewind which worked now correct fileno */
			mt_status.mt_fileno = 0;
			mt_status.mt_blkno = 0;
		}
		if (mt_fileno < mt_status.mt_fileno) {
			mt_com.mt_op = MTNBSF;
			mt_com.mt_count = mt_status.mt_fileno - mt_fileno;
			/* printf("mt: bsf= %d\n", mt_com.mt_count); */
		} else {
			mt_com.mt_op = MTFSF;
			mt_com.mt_count =  mt_fileno - mt_status.mt_fileno;
			/* printf("mt: fsf= %d\n", mt_com.mt_count); */
		}
		if (ioctl(mtfd, MTIOCLTOP, &mt_com) < 0) {
			(void) fprintf(stderr, "%s %s %ld ", tape, comp->c_name,
			    mt_fileno);
			perror("failed");
			return (2);
		}
	} else if (comp->c_code == MTIOCGETDRIVETYPE) {
		return (print_config(mtfd));

	/* Handle regular mag tape ioctls */
	} else if (comp->c_code != MTNOP) {
		int usecnt;

		mt_com.mt_op = comp->c_code;
		usecnt = argc > 2 && comp->c_usecnt;
		mt_com.mt_count = (usecnt ? atoll(argv[2]) : 1);
		if (mt_com.mt_count < 0) {
			(void) fprintf(stderr, "mt: negative %s count\n",
			    comp->c_name);
			return (1);
		}
		if (ioctl(mtfd, MTIOCLTOP, &mt_com) < 0) {
			/*
			 * If we asked for a seek and it returns a tell
			 * we attempted to seek more then there was.
			 */
			if (mt_com.mt_op == MTTELL &&
			    comp->c_code == MTSEEK) {
				(void) printf("partial seek:at block = %llu.\n",
				    mt_com.mt_count);
			} else {
				(void) fprintf(stderr, "%s %s %lld ", tape,
				    comp->c_name, mt_com.mt_count);
				perror("failed");
			}
			return (2);
		}
		if (mt_com.mt_op == MTTELL) {
			(void) printf("At block = %llu.\n", mt_com.mt_count);
		}


	/* Handle status ioctl */
	} else {
		if (ioctl(mtfd, MTIOCGET, (char *)&mt_status) < 0) {
			perror("mt");
			return (2);
		}
		return (status(mtfd, &mt_status));
	}
	return (0);
}

static int
print_config(int mtfd)
{
	struct mtdrivetype mdt;
	struct mtdrivetype_request mdt_req;
	char cfgname[48];
	char tmp[2];
	char *name;
	int i;

	mdt_req.size = sizeof (mdt);
	mdt_req.mtdtp = &mdt;

	if (ioctl(mtfd, MTIOCGETDRIVETYPE, &mdt_req) != 0) {
		perror("mt config");
		return (2);
	}

	/*
	 * remove trailing spaces from product id.
	 */
	for (i = VIDPIDLEN; i; i--) {
		if (isspace(mdt.vid[i]) || mdt.vid[i] == '*') {
			mdt.vid[i] = 0;
		} else if (mdt.vid[i] == 0) {
			continue;
		} else {
			break;
		}
	}

	/*
	 * If this is a generic name display the Vid and Pid instead.
	 */
	if (strstr(mdt.name, "Vendor '") == NULL) {
		name = mdt.name;
	} else {
		name = mdt.vid;
	}

	/*
	 * Attempt to create a configuration name using vid and pid.
	 */
	(void) strcpy(cfgname, "CFG");

	for (tmp[1] = i = 0; i < VIDPIDLEN; i++) {
		if (!isalnum(name[i]))
			continue;
		if (isspace(name[i]))
			continue;
		tmp[0] = toupper(name[i]);
		(void) strncat(cfgname, tmp, 1);
	}

	(void) printf("\"%s\", \"%s\", \"%s\";\n", mdt.vid, name, cfgname);

	/*
	 * Don't want show some bits, ST_DYNAMIC is set in the driver
	 * so one can tell that its not a compiled in config.
	 * The ST_LONG_ERASE and ST_LONG_TIMEOUTS are not displayed
	 * becouse the timeout values below already reflect them being
	 * set.
	 * Also ST_KNOWS_MEDIA is not displayed as it can not be configured
	 * from an st.conf entry.
	 */
	(void) printf("%s = 2,0x%X,%d,0x%X,", cfgname,
	    mdt.type, mdt.bsize, mdt.options &
	    ~(ST_DYNAMIC | ST_LONG_ERASE | ST_LONG_TIMEOUTS | ST_KNOWS_MEDIA));

	(void) printf("4,0x%2.2X,0x%2.2X,0x%2.2X,0x%2.2X,%d,",
	    mdt.densities[0], mdt.densities[1], mdt.densities[2],
	    mdt.densities[3], mdt.default_density >> 3);

	(void) printf("%d,%d,%d,%d,%d,%d,%d;\n", mdt.non_motion_timeout,
	    mdt.io_timeout, mdt.rewind_timeout, mdt.space_timeout,
	    mdt.load_timeout, mdt.unload_timeout, mdt.erase_timeout);

	return (0);
}

/*
 * Interpret the status buffer returned
 */
static int
status(int mtfd, struct mtget *bp)
{
	struct mtdrivetype mdt;
	struct mtdrivetype_request mdt_req;
	const char *name = (char *)NULL;

	/*
	 * Make a call to MTIOCGETDRIVETYPE ioctl, Also use old method
	 * of MT_TAPE_INFO for now, but MT_TAPE_INFO should dissapear in 2.7
	 */
	mdt_req.size = sizeof (struct mtdrivetype);
	mdt_req.mtdtp = &mdt;

	if (ioctl(mtfd, MTIOCGETDRIVETYPE, &mdt_req) == 0) {
		name = mdt.name;
		if (strstr(mdt.name, "Vendor '") != NULL) {
			(void) printf("Unconfigured Drive: ");
		}
	} else {
		perror("mt drivetype");
		return (2);
	}

	/* Handle SCSI tape drives specially. */
	if ((bp->mt_flags & MTF_SCSI)) {
		if (name == (char *)NULL) {
			name = "SCSI";
		}


		(void) printf("%s tape drive:\n", name);

		(void) printf("   sense key(0x%x)= %s   residual= %ld   ",
		    bp->mt_erreg, print_key(bp->mt_erreg), bp->mt_resid);
		(void) printf("retries= %d\n", bp->mt_dsreg);
		/*
		 * Can overflow the signed numbers.
		 * fileno will be -1 on error but all other positions are
		 * positive. blkno will never be negative.
		 */
		if (bp->mt_fileno == -1) {
			(void) printf("   file no= -1   block no= %lu\n",
			    (unsigned long)bp->mt_blkno);
		} else {
			(void) printf("   file no= %lu   block no= %lu\n",
			    (unsigned long)bp->mt_fileno,
			    (unsigned long)bp->mt_blkno);
		}
		if ((bp->mt_flags & MTF_WORM_MEDIA) != 0) {
			(void) printf("   WORM media\n");
		}
	} else {
		/* Handle non-SCSI drives here. */
		if (name == NULL) {
			(void) printf("unknown tape drive type (0x%x)\n",
			    mdt.type);
			return (2);
		}
		(void) printf("%s tape drive:\n   residual= %ld", name,
		    bp->mt_resid);
		printreg("   ds", (ushort_t)bp->mt_dsreg, 0);
		printreg("   er", (ushort_t)bp->mt_erreg, 0);
		(void) putchar('\n');
	}
	return (0);
}


/*
 * Define SCSI sense key error messages.
 *
 * The first 16 sense keys are SCSI standard
 * sense keys. The keys after this are
 * Sun Specifice 'sense' keys- e.g., crap.
 */

static char *sense_keys[] = {
	"No Additional Sense",		/* 0x00 */
	"Soft Error",			/* 0x01 */
	"Not Ready",			/* 0x02 */
	"Media Error",			/* 0x03 */
	"Hardware Error",		/* 0x04 */
	"Illegal Request",		/* 0x05 */
	"Unit Attention",		/* 0x06 */
	"Write Protected",		/* 0x07 */
	"Blank Check",			/* 0x08 */
	"Vendor Unique",		/* 0x09 */
	"Copy Aborted",			/* 0x0a */
	"Aborted Command",		/* 0x0b */
	"Equal Error",			/* 0x0c */
	"Volume Overflow",		/* 0x0d */
	"Miscompare Error",		/* 0x0e */
	"Reserved",			/* 0x0f */
#ifdef	sun
	"fatal",			/* 0x10 */
	"timeout",			/* 0x11 */
	"EOF",				/* 0x12 */
	"EOT",				/* 0x13 */
	"length error",			/* 0x14 */
	"BOT",				/* 0x15 */
	"wrong tape media",		/* 0x16 */
#endif
};

/*
 * Return the text string associated with the sense key value.
 */
static char *
print_key(short key_code)
{
	static char unknown[32];

	if ((key_code >= 0) &&
	    (key_code < (sizeof (sense_keys) / sizeof (sense_keys[0])))) {
		return (sense_keys[key_code]);
	}

	(void) sprintf(unknown, "unknown sense key: 0x%x",
	    (unsigned int) key_code);
	return (unknown);
}


/*
 * Print a register a la the %b format of the kernel's printf
 */
static void
printreg(char *s, ushort_t v, char *bits)
{
	int i, any = 0;
	char c;

	if (bits && *bits == 8) {
		(void) printf("%s = %o", s, v);
	} else {
		(void) printf("%s = %x", s, v);
	}
	bits++;
	if (v && bits) {
		(void) putchar('<');
		while ((i = *bits++) != 0) {
			if (v & (1 << (i-1))) {
				if (any) {
					(void) putchar(',');
				}
				any = 1;
				for (; (c = *bits) > 32; bits++) {
					(void) putchar(c);
				}
			} else {
				for (; *bits > 32; bits++)
					;
			}
		}
		(void) putchar('>');
	}
}
